infrastructure/src/main/java/org/corfudb/infrastructure/log/statetransfer/StateTransferManager.java

```diff
@@ -0,0 +1,277 @@
+    @Getter
+    @ToString
+    @Builder
+    public static class CurrentTransferSegment {
+        /**
+         * Start address of a segment.
+         */
+        @Default
+        private final long startAddress = NON_ADDRESS;
+        /**
+         * End address of a segment.
+         */
+        @Default
+        private final long endAddress = NON_ADDRESS;
+        /**
+         * A future that hold the status of a transfer of a segment.
+         */
+        @Default
+        @NonNull
+        private final CurrentTransferSegmentStatus status;
+
+        /**
+         * If end address and start address are valid, compute the total number of transferred.
+         *
+         * @return Sum of total transferred.
+         */
+        public long computeTotalTransferred() {
+            return endAddress - startAddress + 1L;
+        }
+    }
+
+    /**
+     * A data class that represent a status of a segment to be transferred.
+     */
+    @Getter
+    @ToString
+    @EqualsAndHashCode
+    @Builder
+    public static class CurrentTransferSegmentStatus {
+        /**
+         * A state of a segment.
+         */
+        @Default
+        private final SegmentState segmentState = RESTORED;
+        /**
+         * Total number of records transferred for this segment.
+         */
+        @Default
+        private final long totalTransferred = 0L;
+        /**
+         * An optional cause of failure for this segment.
+         */
+        @Default
+        @Exclude
+        private final Optional<StreamProcessFailure> causeOfFailure = Optional.empty();
+    }
+
+    @Getter
+    @NonNull
+    private final StreamLog streamLog;
+
+    @Getter
+    @NonNull
+    private final int batchSize;
+
+    /**
+     * Given a range, return the addresses that are currently not present in the stream log.
+     *
+     * @param rangeStart Start address.
+     * @param rangeEnd   End address.
+     * @return A list of addresses, currently not present in the stream log.
+     */
+    ImmutableList<Long> getUnknownAddressesInRange(long rangeStart, long rangeEnd) {
+        Set<Long> knownAddresses = streamLog
+                .getKnownAddressesInRange(rangeStart, rangeEnd);
+
+        return LongStream.range(rangeStart, rangeEnd + 1L)
+                .filter(address -> !knownAddresses.contains(address))
+                .boxed()
+                .collect(ImmutableList.toImmutableList());
+    }
+
+    /**
+     * Performs the state transfer for the current transfer segments and also
+     * updates their state as a result.
+     *
+     * @param currentTransferSegments A list of the segment currently present in the system.
+     * @param batchProcessor          An instance of a batch processor.
+     * @return A list with the updated transfer segments.
+     */
+    public ImmutableList<CurrentTransferSegment> handleTransfer
+    (List<CurrentTransferSegment> currentTransferSegments,
+     StateTransferBatchProcessor batchProcessor) {
```

Comment:
It is not very crucial for this implementation, and for the future implementation this would be incorrect since the batch processors will be actually created dynamically by a State Transfer Manager.

Reply:
Question:
Do you consider then creating StateTarnsferManager along with BatchProcessor dynamically?
It should make code more consistent anyway?